src/core/thread/mlr_manager.cpp

```diff
@@ -81,6 +83,141 @@ void MlrManager::HandleBackboneRouterPrimaryUpdate(BackboneRouter::Leader::State
+    for (Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress().IsMulticastLargerThanRealmLocal() && addr.GetMlrState() == kMlrStateToRegister &&
+            IsAddressMlrRegisteredByAnyChild(addr.GetAddress()))
+        {
+            addr.SetMlrState(kMlrStateRegistered);
+        }
+    }
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+    ScheduleSend(0);
+}
+
+bool MlrManager::IsAddressMlrRegisteredByNetif(const Ip6::Address &aAddress)
+{
+    bool ret = false;
+
+    OT_ASSERT(aAddress.IsMulticastLargerThanRealmLocal());
+
+    for (const Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress() == aAddress && addr.GetMlrState() == kMlrStateRegistered)
+        {
+            ExitNow(ret = true);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void MlrManager::SetNetifMulticastAddressMlrState(MlrState aFromState, MlrState aToState)
+{
+    for (Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress().IsMulticastLargerThanRealmLocal() && addr.GetMlrState() == aFromState)
+        {
+            addr.SetMlrState(aToState);
+        }
+    }
+}
+#endif // OPENTHREAD_CONFIG_MLR_ENABLE
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+bool MlrManager::IsAddressMlrRegisteredByAnyChildExcept(const Ip6::Address &aAddress, const Child *aExceptChild)
+{
+    bool ret = false;
+
+    OT_ASSERT(aAddress.IsMulticastLargerThanRealmLocal());
+
+    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
+    {
+        if (&child != aExceptChild && child.HasMlrRegisteredAddress(aAddress))
+        {
+            ExitNow(ret = true);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void MlrManager::UpdateProxiedSubscriptions(Child &             aChild,
+                                            const Ip6::Address *aRegisteredAddresses,
+                                            uint16_t            aRegisteredAddressNum)
+{
+    Ip6::Address address;
+    auto         childAddresses = aChild.IterateIp6Addresses(Ip6::Address::kTypeMulticastLargerThanRealmLocal);
+
+    VerifyOrExit(aChild.IsStateValidOrRestoring(), OT_NOOP);
+
+    // Search the new multicast addresses and set its flag accordingly
+    for (auto iter = childAddresses.begin(); iter != childAddresses.end(); iter++)
+    {
+        bool isMlrRegistered = false;
+
+        if (!address.IsMulticastLargerThanRealmLocal())
+        {
+            continue;
+        }
+
+        // Check if it's a new multicast address against old addresses
+        for (size_t i = 0; i < aRegisteredAddressNum; i++)
+        {
+            if (aRegisteredAddresses[i] == address)
+            {
+                isMlrRegistered = true;
+                break;
+            }
+        }
```

Comment:
Can this work instead?
```
isMlrRegistered = (aChild.GetAddressMlrState(iter.GetAsIndex()) == kMlrStateRegistered);
```

Reply:
No. `aChild` is the current Child whose addresses are updated via `Child Update Requests`. 
This code is executed after `aChild` has updated all its addresses to new address.  So we should check current address against the original MLR registered addresses, rather than current addresses.